src/core/device.c

```diff
@@ -114,12 +159,134 @@ static void device_done(Unit *u) {
+                                       "io_cost_qos is enabled but ctrl is not set, parameters not valid");
+
+        /* Will consider ctrl value of auto to be the same as not enabled.
+         * In both cases the kernel will automatically fill all other io_cost_qos
+         * parameteters so we do not need to input them */
+        if (io_cost_qos->ctrl == IO_COST_CTRL_AUTO || !io_cost_qos->enabled) {
+                log_info("ctrl is set to auto or io_cost_qos is not enabled, setting all other values to kernel default");
+                xsprintf(buf, "%u:%u enable=%d ctrl=%.4s\n",
+                         major(dev),
+                         minor(dev),
+                         io_cost_qos->enabled,
+                         io_cost_ctrl_to_string(IO_COST_CTRL_AUTO));
+
+        } else if (io_cost_qos->ctrl == IO_COST_CTRL_USER) {
+                if (io_cost_qos->min > io_cost_qos->max)
+                         return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                "io_cost_qos min is greater than max, not setting io.cost.qos");
+
+                xsprintf(buf, "%u:%u enable=%d ctrl=%.4s rpct=%u.%02u rlat=%u wpct=%u.%02u wlat=%u min=%u.%02u max=%u.%02u\n",
+                         major(dev),
+                         minor(dev),
+                         io_cost_qos->enabled,
+                         io_cost_ctrl_to_string(io_cost_qos->ctrl),
+                         io_cost_qos->read_latency_percentile / 100,
+                         io_cost_qos->read_latency_percentile % 100,
+                         io_cost_qos->read_latency_threshold,
+                         io_cost_qos->write_latency_percentile / 100,
+                         io_cost_qos->write_latency_percentile % 100,
+                         io_cost_qos->write_latency_threshold,
+                         io_cost_qos->min / 100,
+                         io_cost_qos->min % 100,
+                         io_cost_qos->max / 100,
+                         io_cost_qos->max % 100);
+
+        } else
+                return 0;
+
+        (void) set_attribute_and_warn(root_slice_unit, "io", "io.cost.qos", buf);
+
+        return 0;
+}
+
+static int root_cgroup_apply_block_device_io_cost_model(Unit *root_slice_unit, Device *d) {
+        char buf[DECIMAL_STR_MAX(dev_t)*2+9+12+(9+DECIMAL_STR_MAX(uint64_t)+1)*6];
+        IOCostModel *io_cost_model;
+        dev_t dev;
+        int r;
+
+        assert(root_slice_unit);
+        assert(d);
+
+        io_cost_model = &d->io_cost_model;
+
+        /* If ctrl is invalid it means that io_cost_model is not configured
+         * for this device */
+        if (!io_cost_model || io_cost_model->ctrl == _IO_COST_CTRL_INVALID)
+                return 0;
+
+        if (!verify_io_cost_supported(d))
+                return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "io cost is not supported");
+
+        r = lookup_block_device(d->devname, &dev);
+        if (r < 0)
+                return r;
+
```

Comment:
Can be these common checks (same in `root_cgroup_apply_block_device_io_cost_qos()`) factored out into the caller?

Edit: Ah, it's same what @anitazha says in the caller location.